src/software/ai/hl/stp/tactic/attacker/attacker_tactic.cpp

```diff
@@ -1,111 +1,179 @@
 #include "software/ai/hl/stp/tactic/attacker/attacker_tactic.h"
 
-#include "proto/message_translation/tbots_geometry.h"
-#include "shared/constants.h"
-#include "software/ai/evaluation/calc_best_shot.h"
 #include "software/logger/logger.h"
-#include "software/world/ball.h"
 
-AttackerTactic::AttackerTactic(TbotsProto::AiConfig ai_config)
+AttackerTactic::AttackerTactic(std::shared_ptr<Strategy> strategy)
     : Tactic({RobotCapability::Kick, RobotCapability::Chip, RobotCapability::Move}),
-      fsm_map(),
-      best_pass_so_far(std::nullopt),
-      pass_committed(false),
-      chip_target(std::nullopt),
-      ai_config(ai_config)
+      strategy_(strategy),
+      q_function_(std::make_shared<LinearQFunction<AttackerMdpState, AttackerMdpAction>>(
+          AttackerMdpFeatureExtractor(),
+          strategy_->getAiConfig().attacker_tactic_config().learning_rate(),
+          strategy_->getAiConfig().attacker_tactic_config().discount_factor(),
+          ATTACKER_MDP_Q_FUNCTION_INITIAL_WEIGHTS_FILE)),
+      action_selection_strategy_(
+          std::make_shared<SoftmaxStrategy<AttackerMdpState, AttackerMdpAction>>(
+              strategy_->getAiConfig()
+                  .attacker_tactic_config()
+                  .action_selection_temperature())),
+      policy_(q_function_, action_selection_strategy_),
+      attacker_mdp_reward_function_(strategy_),
+      current_skill_(nullptr)
 {
-    for (RobotId id = 0; id < MAX_ROBOT_IDS; id++)
-    {
-        fsm_map[id] = std::make_unique<FSM<AttackerFSM>>(
-            DribbleFSM(ai_config.dribble_tactic_config()),
-            AttackerFSM(ai_config.attacker_tactic_config()));
-    }
 }
 
-void AttackerTactic::updateControlParams(const Pass& best_pass_so_far,
-                                         bool pass_committed)
+void AttackerTactic::accept(TacticVisitor& visitor) const
 {
-    // Update the control parameters stored by this Tactic
-    this->best_pass_so_far = best_pass_so_far;
-    this->pass_committed   = pass_committed;
+    visitor.visit(*this);
 }
 
-void AttackerTactic::updateControlParams(std::optional<Point> chip_target)
+std::string AttackerTactic::getFSMState() const
 {
-    this->chip_target = chip_target;
+    std::string state = "Unknown";
+    if (current_skill_ && last_execution_robot)
+    {
+        state = current_skill_->getFSMState(*last_execution_robot);
+    }
+    return state;
 }
```

Comment:
When the tactic first starts it will have no robot that last executed it, so this will be false. I can't return an empty string since it causes a sigabrt for some reason